packages/alfa-aria/src/node.ts

```diff
@@ -222,7 +223,70 @@ export namespace Node {
 
     const root = node.root({ flattened: true });
 
-    build(root, device);
+    // Before we start constructing the accessibility tree, we need to resolve
+    // explicit ownership of elements as specified by the `aria-owns` attribute.
+    // https://w3c.github.io/aria/#aria-owns
+
+    // Find all elements in the tree. As explicit ownership is specified via ID
+    // references, it cannot cross shadow or document boundaries.
+    const elements = root.inclusiveDescendants().filter(dom.Element.isElement);
+
+    // Build a map from ID -> element to allow fast resolution of ID references.
+    // The collected references are added to the map in reverse order to ensure
+    // that the first occurrence of a given ID is what ends up in the map in
+    // event of duplicates.
+    const ids = Map.from(
+      elements
+        .collect((element) => element.id.map((id) => [id, element] as const))
+        .reverse()
```

Comment:
Got to point out that this is going through the list twice while the first pass could put stuff to a LIFO accumulator to do the `reverse` at the same time. Not a big deal, I guess.

Reply:
Fortunately, it's not a list; it's a lazy sequence. All `.reverse()` does is therefore force the sequence in reverse order, applying the `#collect()` step at every value encountered. Now, `Map.from()` also needs to force the sequence and, as it's already forced, can simply traverse the existing sequence as a linked list. If we want to avoid both forcing _and_ separately traversing the sequence, `#collect()` could be followed by a `#reduce()` that builds the map and skips duplicate entries:


```ts
const ids = elements
  .collect((element) => element.id.map((id) => [id, element] as const))
  .reduce(
    (map, [id, element]) => (map.has(id) ? map : map.set(id, element)),
    Map.empty<string, dom.Element>()
  );
```

I don't know if any potential benefits outweigh the reduced clarity though 🤔